// plywood/build/external/druidExternal.js
import * as tslib_1 from "tslib";
import * as hasOwnProp from 'has-own-prop';
import { Transform } from 'readable-stream';
import * as toArray from 'stream-to-array';
import { AttributeInfo, Range, Set, TimeRange } from '../datatypes/index';
import { $, CardinalityExpression, ChainableExpression, Expression, InExpression, IsExpression, MatchExpression, MaxExpression, MinExpression, NumberBucketExpression, RefExpression, TimeBucketExpression, TimeFloorExpression, TimePartExpression } from '../expressions/index';
import { dictEqual, ExtendableError, nonEmptyLookup, shallowCopy } from '../helper/utils';
import { External } from './baseExternal';
import { DruidAggregationBuilder } from './utils/druidAggregationBuilder';
import { DruidExtractionFnBuilder } from './utils/druidExtractionFnBuilder';
import { DruidFilterBuilder } from './utils/druidFilterBuilder';
import { DruidHavingFilterBuilder } from './utils/druidHavingFilterBuilder';
var InvalidResultError = (function (_super) {
    tslib_1.__extends(InvalidResultError, _super);
    function InvalidResultError(message, result) {
        var _this = _super.call(this, message) || this;
        _this.result = result;
        return _this;
    }
    return InvalidResultError;
}(ExtendableError));
export { InvalidResultError };
function expressionNeedsNumericSort(ex) {
    var type = ex.type;
    return (type === 'NUMBER' || type === 'NUMBER_RANGE');
}
function simpleJSONEqual(a, b) {
    return JSON.stringify(a) === JSON.stringify(b);
}
var DruidExternal = (function (_super) {
    tslib_1.__extends(DruidExternal, _super);
    function DruidExternal(parameters) {
        var _this = _super.call(this, parameters, dummyObject) || this;
        _this._ensureEngine("druid");
        _this._ensureMinVersion("0.8.0");
        _this.timeAttribute = parameters.timeAttribute || DruidExternal.TIME_ATTRIBUTE;
        _this.customAggregations = parameters.customAggregations;
        _this.customTransforms = parameters.customTransforms;
        _this.allowEternity = parameters.allowEternity;
        _this.allowSelectQueries = parameters.allowSelectQueries;
        var introspectionStrategy = parameters.introspectionStrategy || DruidExternal.DEFAULT_INTROSPECTION_STRATEGY;
        if (DruidExternal.VALID_INTROSPECTION_STRATEGIES.indexOf(introspectionStrategy) === -1) {
            throw new Error("invalid introspectionStrategy '" + introspectionStrategy + "'");
        }
        _this.introspectionStrategy = introspectionStrategy;
        _this.exactResultsOnly = parameters.exactResultsOnly;
        _this.context = parameters.context;
        return _this;
    }
    DruidExternal.fromJS = function (parameters, requester) {
        if (typeof parameters.druidVersion === 'string') {
            parameters.version = parameters.druidVersion;
            console.warn("'druidVersion' parameter is deprecated, use 'version: " + parameters.version + "' instead");
        }
        var value = External.jsToValue(parameters, requester);
        value.timeAttribute = parameters.timeAttribute;
        value.customAggregations = parameters.customAggregations || {};
        value.customTransforms = parameters.customTransforms || {};
        value.allowEternity = Boolean(parameters.allowEternity);
        value.allowSelectQueries = Boolean(parameters.allowSelectQueries);
        value.introspectionStrategy = parameters.introspectionStrategy;
        value.exactResultsOnly = Boolean(parameters.exactResultsOnly);
        value.context = parameters.context;
        return new DruidExternal(value);
    };
    DruidExternal.getSourceList = function (requester) {
        return toArray(requester({ query: { queryType: 'sourceList' } }))
            .then(function (sourcesArray) {
            var sources = sourcesArray[0];
            if (!Array.isArray(sources))
                throw new InvalidResultError('invalid sources response', sources);
            return sources.sort();
        });
    };
    DruidExternal.getVersion = function (requester) {
        return toArray(requester({
            query: {
                queryType: 'status'
            }
        }))
            .then(function (res) {
            return res[0].version;
        });
    };
    DruidExternal.timeBoundaryPostTransformFactory = function (applies) {
        return new Transform({
            objectMode: true,
            transform: function (d, encoding, callback) {
                if (applies) {
                    var datum = {};
                    for (var _i = 0, applies_1 = applies; _i < applies_1.length; _i++) {
                        var apply = applies_1[_i];
                        var name = apply.name;
                        if (typeof d === 'string') {
                            datum[name] = new Date(d);
                        }
                        else {
                            if (apply.expression.op === 'max') {
                                datum[name] = new Date((d['maxIngestedEventTime'] || d['maxTime']));
                            }
                            else {
                                datum[name] = new Date(d['minTime']);
                            }
                        }
                    }
                    callback(null, {
                        type: 'datum',
                        datum: datum
                    });
                }
                else {
                    callback(null, {
                        type: 'value',
                        value: new Date((d['maxIngestedEventTime'] || d['maxTime'] || d['minTime']))
                    });
                }
            }
        });
    };
    DruidExternal.selectNextFactory = function (limit, descending) {
        var resultsSoFar = 0;
        return function (prevQuery, prevResultLength, prevMeta) {
            if (prevResultLength === 0)
                return null;
            var pagingIdentifiers = prevMeta.pagingIdentifiers;
            if (prevResultLength < prevQuery.pagingSpec.threshold)
                return null;
            resultsSoFar += prevResultLength;
            if (resultsSoFar >= limit)
                return null;
            pagingIdentifiers = DruidExternal.movePagingIdentifiers(pagingIdentifiers, descending ? -1 : 1);
            prevQuery.pagingSpec.pagingIdentifiers = pagingIdentifiers;
            prevQuery.pagingSpec.fromNext = false;
            prevQuery.pagingSpec.threshold = Math.min(limit - resultsSoFar, DruidExternal.SELECT_MAX_LIMIT);
            return prevQuery;
        };
    };
    DruidExternal.generateMaker = function (aggregation) {
        if (!aggregation)
            return null;
        var type = aggregation.type, fieldName = aggregation.fieldName;
        if (type === 'longSum' && fieldName === 'count') {
            return Expression._.count();
        }
        if (!fieldName) {
            var fieldNames = aggregation.fieldNames;
            if (!Array.isArray(fieldNames) || fieldNames.length !== 1)
                return null;
            fieldName = fieldNames[0];
        }
        var expression = $(fieldName);
        switch (type) {
            case "count":
                return Expression._.count();
            case "doubleSum":
            case "longSum":
                return Expression._.sum(expression);
            case "javascript":
                var fnAggregate = aggregation.fnAggregate, fnCombine = aggregation.fnCombine;
                if (fnAggregate !== fnCombine || fnCombine.indexOf('+') === -1)
                    return null;
                return Expression._.sum(expression);
            case "doubleMin":
            case "longMin":
                return Expression._.min(expression);
            case "doubleMax":
            case "longMax":
                return Expression._.max(expression);
            default:
                return null;
        }
    };
    DruidExternal.columnMetadataToRange = function (columnMetadata) {
        var minValue = columnMetadata.minValue, maxValue = columnMetadata.maxValue;
        if (minValue == null || maxValue == null)
            return null;
        return Range.fromJS({
            start: minValue,
            end: maxValue,
            bounds: '[]'
        });
    };
    DruidExternal.segmentMetadataPostProcess = function (timeAttribute, numericColumnSupport, res) {
        var res0 = res[0];
        if (!res0 || !res0.columns)
            throw new InvalidResultError('malformed segmentMetadata response', res);
        var columns = res0.columns;
        var aggregators = res0.aggregators || {};
        var foundTime = false;
        var attributes = [];
        for (var name in columns) {
            if (!hasOwnProp(columns, name))
                continue;
            var columnData = columns[name];
            if (columnData.errorMessage || columnData.size < 0)
                continue;
            if (name === DruidExternal.TIME_ATTRIBUTE) {
                attributes.unshift(new AttributeInfo({
                    name: timeAttribute,
                    type: 'TIME',
                    nativeType: '__time',
                    cardinality: columnData.cardinality,
                    range: DruidExternal.columnMetadataToRange(columnData)
                }));
                foundTime = true;
            }
            else {
                if (name === timeAttribute)
                    continue;
                var nativeType = columnData.type;
                switch (columnData.type) {
                    case 'DOUBLE':
                    case 'FLOAT':
                    case 'LONG':
                        attributes.push(new AttributeInfo({
                            name: name,
                            type: 'NUMBER',
                            nativeType: nativeType,
                            unsplitable: !(numericColumnSupport && !hasOwnProp(aggregators, name)),
                            maker: DruidExternal.generateMaker(aggregators[name]),
                            cardinality: columnData.cardinality,
                            range: DruidExternal.columnMetadataToRange(columnData)
                        }));
                        break;
                    case 'STRING':
                        attributes.push(new AttributeInfo({
                            name: name,
                            type: columnData.hasMultipleValues ? 'SET/STRING' : 'STRING',
                            nativeType: nativeType,
                            cardinality: columnData.cardinality,
                            range: DruidExternal.columnMetadataToRange(columnData)
                        }));
                        break;
                    case 'hyperUnique':
                    case 'approximateHistogram':
                    case 'thetaSketch':
                        attributes.push(new AttributeInfo({
                            name: name,
                            type: 'NULL',
                            nativeType: nativeType,
                            unsplitable: true
                        }));
                        break;
                    default:
                        attributes.push(new AttributeInfo({
                            name: name,
                            type: 'NULL',
                            nativeType: nativeType
                        }));
                        break;
                }
            }
        }
        if (!foundTime)
            throw new Error("no valid " + DruidExternal.TIME_ATTRIBUTE + " in segmentMetadata response");
        return attributes;
    };
    DruidExternal.introspectPostProcessFactory = function (timeAttribute, res) {
        var res0 = res[0];
        if (!Array.isArray(res0.dimensions) || !Array.isArray(res0.metrics)) {
            throw new InvalidResultError('malformed GET introspect response', res);
        }
        var attributes = [
            new AttributeInfo({ name: timeAttribute, type: 'TIME', nativeType: '__time' })
        ];
        res0.dimensions.forEach(function (dimension) {
            if (dimension === timeAttribute)
                return;
            attributes.push(new AttributeInfo({ name: dimension, type: 'STRING', nativeType: 'STRING' }));
        });
        res0.metrics.forEach(function (metric) {
            if (metric === timeAttribute)
                return;
            attributes.push(new AttributeInfo({ name: metric, type: 'NUMBER', nativeType: 'FLOAT', unsplitable: true }));
        });
        return attributes;
    };
    DruidExternal.movePagingIdentifiers = function (pagingIdentifiers, increment) {
        var newPagingIdentifiers = {};
        for (var key in pagingIdentifiers) {
            if (!hasOwnProp(pagingIdentifiers, key))
                continue;
            newPagingIdentifiers[key] = pagingIdentifiers[key] + increment;
        }
        return newPagingIdentifiers;
    };
    DruidExternal.prototype.valueOf = function () {
        var value = _super.prototype.valueOf.call(this);
        value.timeAttribute = this.timeAttribute;
        value.customAggregations = this.customAggregations;
        value.customTransforms = this.customTransforms;
        value.allowEternity = this.allowEternity;
        value.allowSelectQueries = this.allowSelectQueries;
        value.introspectionStrategy = this.introspectionStrategy;
        value.exactResultsOnly = this.exactResultsOnly;
        value.context = this.context;
        return value;
    };
    DruidExternal.prototype.toJS = function () {
        var js = _super.prototype.toJS.call(this);
        if (this.timeAttribute !== DruidExternal.TIME_ATTRIBUTE)
            js.timeAttribute = this.timeAttribute;
        if (nonEmptyLookup(this.customAggregations))
            js.customAggregations = this.customAggregations;
        if (nonEmptyLookup(this.customTransforms))
            js.customTransforms = this.customTransforms;
        if (this.allowEternity)
            js.allowEternity = true;
        if (this.allowSelectQueries)
            js.allowSelectQueries = true;
        if (this.introspectionStrategy !== DruidExternal.DEFAULT_INTROSPECTION_STRATEGY)
            js.introspectionStrategy = this.introspectionStrategy;
        if (this.exactResultsOnly)
            js.exactResultsOnly = true;
        if (this.context)
            js.context = this.context;
        return js;
    };
    DruidExternal.prototype.equals = function (other) {
        return _super.prototype.equals.call(this, other) &&
            this.timeAttribute === other.timeAttribute &&
            simpleJSONEqual(this.customAggregations, other.customAggregations) &&
            simpleJSONEqual(this.customTransforms, other.customTransforms) &&
            this.allowEternity === other.allowEternity &&
            this.allowSelectQueries === other.allowSelectQueries &&
            this.introspectionStrategy === other.introspectionStrategy &&
            this.exactResultsOnly === other.exactResultsOnly &&
            dictEqual(this.context, other.context);
    };
    DruidExternal.prototype.canHandleTotal = function () {
        return true;
    };
    DruidExternal.prototype.canHandleFilter = function (filter) {
        return !filter.expression.some(function (ex) { return ex.isOp('cardinality') ? true : null; });
    };
    DruidExternal.prototype.canHandleSplit = function (split) {
        return true;
    };
    DruidExternal.prototype.canHandleSplitExpression = function (ex) {
        return true;
    };
    DruidExternal.prototype.canHandleApply = function (apply) {
        return true;
    };
    DruidExternal.prototype.canHandleSort = function (sort) {
        if (this.isTimeseries()) {
            if (sort.direction !== 'ascending')
                return false;
            return sort.refName() === this.split.firstSplitName();
        }
        else if (this.mode === 'raw') {
            if (sort.refName() !== this.timeAttribute)
                return false;
            if (this.versionBefore('0.9.0'))
                return sort.direction === 'ascending';
            return true;
        }
        else {
            return true;
        }
    };
    DruidExternal.prototype.canHandleLimit = function (limit) {
        return !this.isTimeseries();
    };
    DruidExternal.prototype.canHandleHavingFilter = function (havingFilter) {
        return !this.limit;
    };
    DruidExternal.prototype.isTimeseries = function () {
        var split = this.split;
        if (!split || split.isMultiSplit())
            return false;
        var splitExpression = split.firstSplitExpression();
        if (this.isTimeRef(splitExpression))
            return true;
        var op = splitExpression.op;
        return op === 'timeBucket' || op === 'timeFloor';
    };
    DruidExternal.prototype.getDruidDataSource = function () {
        var source = this.source;
        if (Array.isArray(source)) {
            return {
                type: "union",
                dataSources: source
            };
        }
        else {
            return source;
        }
    };
    DruidExternal.prototype.isTimeRef = function (ex) {
        return ex instanceof RefExpression && ex.name === this.timeAttribute;
    };
    DruidExternal.prototype.splitExpressionToGranularityInflater = function (splitExpression, label) {
        if (this.isTimeRef(splitExpression)) {
            return {
                granularity: 'none',
                inflater: External.timeInflaterFactory(label)
            };
        }
        else if (splitExpression instanceof TimeBucketExpression || splitExpression instanceof TimeFloorExpression) {
            var op = splitExpression.op, operand = splitExpression.operand, duration = splitExpression.duration;
            var timezone = splitExpression.getTimezone();
            if (this.isTimeRef(operand)) {
                return {
                    granularity: {
                        type: "period",
                        period: duration.toString(),
                        timeZone: timezone.toString()
                    },
                    inflater: op === 'timeBucket' ?
                        External.timeRangeInflaterFactory(label, duration, timezone) :
                        External.timeInflaterFactory(label)
                };
            }
        }
        return null;
    };
    DruidExternal.prototype.makeOutputName = function (name) {
        if (name.indexOf('__') === 0) {
            return '***' + name;
        }
        return name;
    };
    DruidExternal.prototype.expressionToDimensionInflater = function (expression, label) {
        var freeReferences = expression.getFreeReferences();
        if (freeReferences.length === 0) {
            var someDim = DruidExternal.TIME_ATTRIBUTE;
            if (this.versionBefore('0.9.2')) {
                var someAttr = this.attributes.filter(function (a) { return a.type === 'STRING'; })[0];
                if (!someAttr)
                    throw new Error("could not find any dimension");
                someDim = someAttr.name;
            }
            return {
                dimension: {
                    type: "extraction",
                    dimension: someDim,
                    outputName: this.makeOutputName(label),
                    extractionFn: new DruidExtractionFnBuilder(this).expressionToExtractionFn(expression)
                },
                inflater: null
            };
        }
        if (freeReferences.length > 1) {
            throw new Error("must have at most 1 reference (has " + freeReferences.length + "): " + expression);
        }
        var referenceName = freeReferences[0];
        var attributeInfo = this.getAttributesInfo(referenceName);
        if (attributeInfo.unsplitable) {
            throw new Error("can not convert " + expression + " to split because it references an un-splitable metric '" + referenceName + "' which is most likely rolled up.");
        }
        var extractionFn = new DruidExtractionFnBuilder(this).expressionToExtractionFn(expression);
        var simpleInflater = External.getSimpleInflater(expression.type, label);
        var dimension = {
            type: "default",
            dimension: attributeInfo.name === this.timeAttribute ? DruidExternal.TIME_ATTRIBUTE : attributeInfo.name,
            outputName: this.makeOutputName(label)
        };
        if (extractionFn) {
            dimension.type = "extraction";
            dimension.extractionFn = extractionFn;
        }
        if (expression instanceof RefExpression) {
            return {
                dimension: dimension,
                inflater: simpleInflater
            };
        }
        if (expression instanceof TimeBucketExpression) {
            return {
                dimension: dimension,
                inflater: External.timeRangeInflaterFactory(label, expression.duration, expression.getTimezone())
            };
        }
        if (expression instanceof TimePartExpression) {
            return {
                dimension: dimension,
                inflater: simpleInflater
            };
        }
        if (expression instanceof NumberBucketExpression) {
            return {
                dimension: dimension,
                inflater: External.numberRangeInflaterFactory(label, expression.size)
            };
        }
        if (expression instanceof CardinalityExpression) {
            return {
                dimension: dimension,
                inflater: External.setCardinalityInflaterFactory(label)
            };
        }
        var effectiveType = Set.unwrapSetType(expression.type);
        if (simpleInflater || effectiveType === 'STRING' || effectiveType === 'NULL') {
            return {
                dimension: dimension,
                inflater: simpleInflater
            };
        }
        throw new Error("could not convert " + expression + " to a Druid dimension");
    };
    DruidExternal.prototype.expressionToDimensionInflaterHaving = function (expression, label, havingFilter) {
        var dimensionInflater = this.expressionToDimensionInflater(expression, label);
        dimensionInflater.having = havingFilter;
        if (expression.type !== 'SET/STRING')
            return dimensionInflater;
        var _a = havingFilter.extractFromAnd(function (hf) {
            if (hf instanceof ChainableExpression) {
                var hfOp = hf.op;
                var hfOperand = hf.operand;
                if (hfOperand instanceof RefExpression && hfOperand.name === label) {
                    if (hfOp === 'match')
                        return true;
                    if (hfOp === 'is')
                        return hf.expression.isOp('literal');
                }
            }
            return false;
        }), extract = _a.extract, rest = _a.rest;
        if (extract.equals(Expression.TRUE))
            return dimensionInflater;
        if (extract instanceof MatchExpression) {
            return {
                dimension: {
                    type: "regexFiltered",
                    delegate: dimensionInflater.dimension,
                    pattern: extract.regexp
                },
                inflater: dimensionInflater.inflater,
                having: rest
            };
        }
        else if (extract instanceof IsExpression) {
            return {
                dimension: {
                    type: "listFiltered",
                    delegate: dimensionInflater.dimension,
                    values: [extract.expression.getLiteralValue()]
                },
                inflater: dimensionInflater.inflater,
                having: rest
            };
        }
        else if (extract instanceof InExpression) {
            return {
                dimension: {
                    type: "listFiltered",
                    delegate: dimensionInflater.dimension,
                    values: extract.expression.getLiteralValue().elements
                },
                inflater: dimensionInflater.inflater,
                having: rest
            };
        }
        return dimensionInflater;
    };
    DruidExternal.prototype.splitToDruid = function (split) {
        var _this = this;
        var leftoverHavingFilter = this.havingFilter;
        var selectedAttributes = this.getSelectedAttributes();
        if (split.isMultiSplit()) {
            var timestampLabel_1 = null;
            var granularity_1 = null;
            var dimensions_1 = [];
            var inflaters_1 = [];
            split.mapSplits(function (name, expression) {
                if (!granularity_1 && !_this.limit && !_this.sort) {
                    var granularityInflater_1 = _this.splitExpressionToGranularityInflater(expression, name);
                    if (granularityInflater_1) {
                        timestampLabel_1 = name;
                        granularity_1 = granularityInflater_1.granularity;
                        inflaters_1.push(granularityInflater_1.inflater);
                        return;
                    }
                }
                var _a = _this.expressionToDimensionInflaterHaving(expression, name, leftoverHavingFilter), dimension = _a.dimension, inflater = _a.inflater, having = _a.having;
                leftoverHavingFilter = having;
                dimensions_1.push(dimension);
                if (inflater) {
                    inflaters_1.push(inflater);
                }
            });
            return {
                queryType: 'groupBy',
                dimensions: dimensions_1,
                timestampLabel: timestampLabel_1,
                granularity: granularity_1 || 'all',
                leftoverHavingFilter: leftoverHavingFilter,
                postTransform: External.postTransformFactory(inflaters_1, selectedAttributes, split.mapSplits(function (name) { return name; }), null)
            };
        }
        var splitExpression = split.firstSplitExpression();
        var label = split.firstSplitName();
        var granularityInflater = this.splitExpressionToGranularityInflater(splitExpression, label);
        if (granularityInflater) {
            return {
                queryType: 'timeseries',
                granularity: granularityInflater.granularity,
                leftoverHavingFilter: leftoverHavingFilter,
                timestampLabel: label,
                postTransform: External.postTransformFactory([granularityInflater.inflater], selectedAttributes, [label], null)
            };
        }
        var dimensionInflater = this.expressionToDimensionInflaterHaving(splitExpression, label, leftoverHavingFilter);
        leftoverHavingFilter = dimensionInflater.having;
        var inflaters = [dimensionInflater.inflater].filter(Boolean);
        if (leftoverHavingFilter.equals(Expression.TRUE) &&
            (this.limit || split.maxBucketNumber() < 1000) &&
            !this.exactResultsOnly) {
            return {
                queryType: 'topN',
                dimension: dimensionInflater.dimension,
                granularity: 'all',
                leftoverHavingFilter: leftoverHavingFilter,
                timestampLabel: null,
                postTransform: External.postTransformFactory(inflaters, selectedAttributes, [label], null)
            };
        }
        return {
            queryType: 'groupBy',
            dimensions: [dimensionInflater.dimension],
            granularity: 'all',
            leftoverHavingFilter: leftoverHavingFilter,
            timestampLabel: null,
            postTransform: External.postTransformFactory(inflaters, selectedAttributes, [label], null)
        };
    };
    DruidExternal.prototype.isMinMaxTimeExpression = function (applyExpression) {
        if (applyExpression instanceof MinExpression || applyExpression instanceof MaxExpression) {
            return this.isTimeRef(applyExpression.expression);
        }
        else {
            return false;
        }
    };
    DruidExternal.prototype.getTimeBoundaryQueryAndPostTransform = function () {
        var _a = this, mode = _a.mode, context = _a.context;
        var druidQuery = {
            queryType: "timeBoundary",
            dataSource: this.getDruidDataSource()
        };
        if (context) {
            druidQuery.context = context;
        }
        var applies = null;
        if (mode === 'total') {
            applies = this.applies;
            if (applies.length === 1) {
                var loneApplyExpression = applies[0].expression;
                druidQuery.bound = loneApplyExpression.op + "Time";
            }
        }
        else if (mode === 'value') {
            var valueExpression = this.valueExpression;
            druidQuery.bound = valueExpression.op + "Time";
        }
        else {
            throw new Error("invalid mode '" + mode + "' for timeBoundary");
        }
        return {
            query: druidQuery,
            context: { timestamp: null },
            postTransform: DruidExternal.timeBoundaryPostTransformFactory(applies)
        };
    };
    DruidExternal.prototype.getQueryAndPostTransform = function () {
        var _this = this;
        var _a = this, mode = _a.mode, applies = _a.applies, sort = _a.sort, limit = _a.limit, context = _a.context;
        if (mode === 'total' && applies && applies.length && applies.every(function (apply) { return _this.isMinMaxTimeExpression(apply.expression); })) {
            return this.getTimeBoundaryQueryAndPostTransform();
        }
        else if (mode === 'value' && this.isMinMaxTimeExpression(this.valueExpression)) {
            return this.getTimeBoundaryQueryAndPostTransform();
        }
        var druidQuery = {
            queryType: 'timeseries',
            dataSource: this.getDruidDataSource(),
            intervals: null,
            granularity: 'all'
        };
        var requesterContext = {
            timestamp: null,
            ignorePrefix: '!',
            dummyPrefix: '***'
        };
        if (context) {
            druidQuery.context = shallowCopy(context);
        }
        var filterAndIntervals = new DruidFilterBuilder(this).filterToDruid(this.getQueryFilter());
        druidQuery.intervals = filterAndIntervals.intervals;
        if (filterAndIntervals.filter) {
            druidQuery.filter = filterAndIntervals.filter;
        }
        var aggregationsAndPostAggregations;
        switch (mode) {
            case 'raw':
                if (!this.allowSelectQueries) {
                    throw new Error("to issues 'select' queries allowSelectQueries flag must be set");
                }
                var selectDimensions_1 = [];
                var selectMetrics_1 = [];
                var inflaters_2 = [];
                var timeAttribute_1 = this.timeAttribute;
                var derivedAttributes_1 = this.derivedAttributes;
                var selectedAttributes = this.getSelectedAttributes();
                selectedAttributes.forEach(function (attribute) {
                    var name = attribute.name, type = attribute.type, nativeType = attribute.nativeType, unsplitable = attribute.unsplitable;
                    if (name === timeAttribute_1) {
                        requesterContext.timestamp = name;
                    }
                    else {
                        if (nativeType === 'STRING' || (!nativeType && !unsplitable)) {
                            var derivedAttribute = derivedAttributes_1[name];
                            if (derivedAttribute) {
                                if (_this.versionBefore('0.9.1')) {
                                    throw new Error("can not have derived attributes in Druid select in " + _this.version + ", upgrade to 0.9.1");
                                }
                                var dimensionInflater = _this.expressionToDimensionInflater(derivedAttribute, name);
                                selectDimensions_1.push(dimensionInflater.dimension);
                                if (dimensionInflater.inflater)
                                    inflaters_2.push(dimensionInflater.inflater);
                                return;
                            }
                            else {
                                selectDimensions_1.push(name);
                            }
                        }
                        else {
                            selectMetrics_1.push(name);
                        }
                    }
                    switch (type) {
                        case 'BOOLEAN':
                            inflaters_2.push(External.booleanInflaterFactory(name));
                            break;
                        case 'NUMBER':
                            inflaters_2.push(External.numberInflaterFactory(name));
                            break;
                        case 'TIME':
                            inflaters_2.push(External.timeInflaterFactory(name));
                            break;
                        case 'SET/STRING':
                            inflaters_2.push(External.setStringInflaterFactory(name));
                            break;
                    }
                });
                if (!selectDimensions_1.length)
                    selectDimensions_1.push(DruidExternal.DUMMY_NAME);
                if (!selectMetrics_1.length)
                    selectMetrics_1.push(DruidExternal.DUMMY_NAME);
                var resultLimit = limit ? limit.value : Infinity;
                druidQuery.queryType = 'select';
                druidQuery.dimensions = selectDimensions_1;
                druidQuery.metrics = selectMetrics_1;
                druidQuery.pagingSpec = {
                    "pagingIdentifiers": {},
                    "threshold": Math.min(resultLimit, DruidExternal.SELECT_INIT_LIMIT)
                };
                var descending = sort && sort.direction === 'descending';
                if (descending) {
                    druidQuery.descending = true;
                }
                return {
                    query: druidQuery,
                    context: requesterContext,
                    postTransform: External.postTransformFactory(inflaters_2, selectedAttributes.map(function (a) { return a.dropOriginInfo(); }), null, null),
                    next: DruidExternal.selectNextFactory(resultLimit, descending)
                };
            case 'value':
                aggregationsAndPostAggregations = new DruidAggregationBuilder(this).makeAggregationsAndPostAggregations([this.toValueApply()]);
                if (aggregationsAndPostAggregations.aggregations.length) {
                    druidQuery.aggregations = aggregationsAndPostAggregations.aggregations;
                }
                if (aggregationsAndPostAggregations.postAggregations.length) {
                    druidQuery.postAggregations = aggregationsAndPostAggregations.postAggregations;
                }
                return {
                    query: druidQuery,
                    context: requesterContext,
                    postTransform: External.valuePostTransformFactory()
                };
            case 'total':
                aggregationsAndPostAggregations = new DruidAggregationBuilder(this).makeAggregationsAndPostAggregations(this.applies);
                if (aggregationsAndPostAggregations.aggregations.length) {
                    druidQuery.aggregations = aggregationsAndPostAggregations.aggregations;
                }
                if (aggregationsAndPostAggregations.postAggregations.length) {
                    druidQuery.postAggregations = aggregationsAndPostAggregations.postAggregations;
                }
                return {
                    query: druidQuery,
                    context: requesterContext,
                    postTransform: External.postTransformFactory([], this.getSelectedAttributes(), [], applies)
                };
            case 'split':
                var split = this.getQuerySplit();
                var splitSpec = this.splitToDruid(split);
                druidQuery.queryType = splitSpec.queryType;
                druidQuery.granularity = splitSpec.granularity;
                if (splitSpec.dimension)
                    druidQuery.dimension = splitSpec.dimension;
                if (splitSpec.dimensions)
                    druidQuery.dimensions = splitSpec.dimensions;
                var leftoverHavingFilter = splitSpec.leftoverHavingFilter;
                var timestampLabel_2 = splitSpec.timestampLabel;
                requesterContext.timestamp = timestampLabel_2;
                var postTransform = splitSpec.postTransform;
                aggregationsAndPostAggregations = new DruidAggregationBuilder(this).makeAggregationsAndPostAggregations(applies);
                if (aggregationsAndPostAggregations.aggregations.length) {
                    druidQuery.aggregations = aggregationsAndPostAggregations.aggregations;
                }
                else if (this.versionBefore('0.9.2')) {
                    druidQuery.aggregations = [{ name: DruidExternal.DUMMY_NAME, type: "count" }];
                }
                if (aggregationsAndPostAggregations.postAggregations.length) {
                    druidQuery.postAggregations = aggregationsAndPostAggregations.postAggregations;
                }
                switch (druidQuery.queryType) {
                    case 'timeseries':
                        if (sort && (sort.direction !== 'ascending' || !split.hasKey(sort.refName()))) {
                            throw new Error('can not sort within timeseries query');
                        }
                        if (limit) {
                            throw new Error('can not limit within timeseries query');
                        }
                        if (!druidQuery.context || !hasOwnProp(druidQuery.context, 'skipEmptyBuckets')) {
                            druidQuery.context = druidQuery.context || {};
                            druidQuery.context.skipEmptyBuckets = "true";
                        }
                        break;
                    case 'topN':
                        var metric = void 0;
                        if (sort) {
                            var inverted = void 0;
                            if (this.sortOnLabel()) {
                                if (expressionNeedsNumericSort(split.firstSplitExpression())) {
                                    if (this.versionBefore('0.9.2')) {
                                        metric = { type: 'alphaNumeric' };
                                    }
                                    else {
                                        metric = { type: 'dimension', ordering: 'numeric' };
                                    }
                                }
                                else {
                                    if (this.versionBefore('0.9.2')) {
                                        metric = { type: 'lexicographic' };
                                    }
                                    else {
                                        metric = { type: 'dimension', ordering: 'lexicographic' };
                                    }
                                }
                                inverted = sort.direction === 'descending';
                            }
                            else {
                                metric = sort.refName();
                                inverted = sort.direction === 'ascending';
                            }
                            if (inverted) {
                                metric = { type: "inverted", metric: metric };
                            }
                        }
                        else {
                            if (this.versionBefore('0.9.2')) {
                                metric = { type: 'lexicographic' };
                            }
                            else {
                                metric = { type: 'dimension', ordering: 'lexicographic' };
                            }
                        }
                        druidQuery.metric = metric;
                        druidQuery.threshold = limit ? limit.value : 1000;
                        break;
                    case 'groupBy':
                        var orderByColumn = null;
                        if (sort) {
                            var col = sort.refName();
                            orderByColumn = {
                                dimension: this.makeOutputName(col),
                                direction: sort.direction
                            };
                            if (this.sortOnLabel()) {
                                if (expressionNeedsNumericSort(split.splits[col])) {
                                    orderByColumn.dimensionOrder = this.versionBefore('0.9.2') ? 'alphanumeric' : 'numeric';
                                }
                            }
                        }
                        else {
                            var splitKeys = split.keys.filter(function (k) { return k !== timestampLabel_2; });
                            if (!splitKeys.length)
                                throw new Error('could not find order by column for group by');
                            var splitKey = splitKeys[0];
                            var keyExpression = split.splits[splitKey];
                            orderByColumn = {
                                dimension: this.makeOutputName(splitKey)
                            };
                            if (expressionNeedsNumericSort(keyExpression)) {
                                orderByColumn.dimensionOrder = this.versionBefore('0.9.2') ? 'alphanumeric' : 'numeric';
                            }
                        }
                        druidQuery.limitSpec = {
                            type: "default",
                            columns: [orderByColumn || this.makeOutputName(split.firstSplitName())]
                        };
                        if (limit) {
                            druidQuery.limitSpec.limit = limit.value;
                        }
                        if (!leftoverHavingFilter.equals(Expression.TRUE)) {
                            druidQuery.having = new DruidHavingFilterBuilder(this).filterToHavingFilter(leftoverHavingFilter);
                        }
                        break;
                }
                return {
                    query: druidQuery,
                    context: requesterContext,
                    postTransform: postTransform
                };
            default:
                throw new Error("can not get query for: " + this.mode);
        }
    };
    DruidExternal.prototype.getIntrospectAttributesWithSegmentMetadata = function (deep) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, requester, timeAttribute, context, analysisTypes, query, res, attributes, resTB, resTB0;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this, requester = _a.requester, timeAttribute = _a.timeAttribute, context = _a.context;
                        analysisTypes = ['aggregators'];
                        if (deep) {
                            analysisTypes.push('cardinality', 'minmax');
                        }
                        query = {
                            queryType: 'segmentMetadata',
                            dataSource: this.getDruidDataSource(),
                            merge: true,
                            analysisTypes: analysisTypes,
                            lenientAggregatorMerge: true
                        };
                        if (context) {
                            query.context = context;
                        }
                        if (this.versionBefore('0.9.0')) {
                            query.analysisTypes = [];
                            delete query.lenientAggregatorMerge;
                        }
                        if (this.versionBefore('0.9.2') && query.dataSource.type === 'union') {
                            query.dataSource = query.dataSource.dataSources[0];
                        }
                        return [4, toArray(requester({ query: query }))];
                    case 1:
                        res = _b.sent();
                        attributes = DruidExternal.segmentMetadataPostProcess(timeAttribute, !this.versionBefore('0.10.0'), res);
                        if (!(deep && attributes.length && !attributes[0].range)) return [3, 3];
                        query = {
                            queryType: "timeBoundary",
                            dataSource: this.getDruidDataSource()
                        };
                        if (context) {
                            query.context = context;
                        }
                        return [4, toArray(requester({ query: query }))];
                    case 2:
                        resTB = _b.sent();
                        resTB0 = resTB[0];
                        attributes[0] = attributes[0].changeRange(TimeRange.fromJS({
                            start: resTB0.minTime,
                            end: resTB0.maxTime,
                            bounds: '[]'
                        }));
                        _b.label = 3;
                    case 3: return [2, attributes];
                }
            });
        });
    };
    DruidExternal.prototype.getIntrospectAttributesWithGet = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, requester, timeAttribute, res;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this, requester = _a.requester, timeAttribute = _a.timeAttribute;
                        return [4, toArray(requester({
                                query: {
                                    queryType: 'introspect',
                                    dataSource: this.getDruidDataSource()
                                }
                            }))];
                    case 1:
                        res = _b.sent();
                        return [2, DruidExternal.introspectPostProcessFactory(timeAttribute, res)];
                }
            });
        });
    };
    DruidExternal.prototype.getIntrospectAttributes = function (deep) {
        var _this = this;
        switch (this.introspectionStrategy) {
            case 'segment-metadata-fallback':
                return this.getIntrospectAttributesWithSegmentMetadata(deep)
                    .catch(function (err) {
                    if (err.message.indexOf("querySegmentSpec can't be null") === -1)
                        throw err;
                    return _this.getIntrospectAttributesWithGet();
                });
            case 'segment-metadata-only':
                return this.getIntrospectAttributesWithSegmentMetadata(deep);
            case 'datasource-get':
                return this.getIntrospectAttributesWithGet();
            default:
                throw new Error('invalid introspectionStrategy');
        }
    };
    DruidExternal.engine = 'druid';
    DruidExternal.type = 'DATASET';
    DruidExternal.DUMMY_NAME = '!DUMMY';
    DruidExternal.TIME_ATTRIBUTE = '__time';
    DruidExternal.VALID_INTROSPECTION_STRATEGIES = ['segment-metadata-fallback', 'segment-metadata-only', 'datasource-get'];
    DruidExternal.DEFAULT_INTROSPECTION_STRATEGY = 'segment-metadata-fallback';
    DruidExternal.SELECT_INIT_LIMIT = 50;
    DruidExternal.SELECT_MAX_LIMIT = 10000;
    return DruidExternal;
}(External));
export { DruidExternal };
External.register(DruidExternal);
