// plywood-druid-requester/build/druidRequester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var plywood_base_api_1 = require("plywood-base-api");
var readable_stream_1 = require("readable-stream");
var request = require("request");
var hasOwnProperty = require("has-own-prop");
var requestPromise = require("request-promise-any");
var concat = require("concat-stream");
var PlainAgent = require("socks5-http-client/lib/Agent");
var SecureAgent = require("socks5-https-client/lib/Agent");
var Combo = require("stream-json/Combo");
var rowBuilder_1 = require("./rowBuilder");
function getDataSourcesFromQuery(query) {
    var queryDataSource = query.dataSource;
    if (!queryDataSource)
        return [];
    if (typeof queryDataSource === 'string') {
        return [queryDataSource];
    }
    else if (queryDataSource.type === "union") {
        return queryDataSource.dataSources;
    }
    else {
        throw new Error("unsupported datasource type '" + queryDataSource.type + "'");
    }
}
function basicUrlBuilder(location, secure) {
    var s = '';
    var defaultPort = 8082;
    if (secure) {
        s = 's';
        defaultPort = 8282;
    }
    return "http" + s + "://" + location.hostname + ":" + (location.port || defaultPort);
}
function druidRequesterFactory(parameters) {
    var locator = parameters.locator, host = parameters.host, timeout = parameters.timeout, protocol = parameters.protocol, ca = parameters.ca, urlBuilder = parameters.urlBuilder, requestDecorator = parameters.requestDecorator, authToken = parameters.authToken, socksHost = parameters.socksHost;
    if (!protocol)
        protocol = 'plain';
    var secure = protocol === 'tls' || protocol === 'tls-loose';
    if (!locator) {
        if (!host)
            throw new Error("must have a `host` or a `locator`");
        locator = plywood_base_api_1.basicLocator(host, secure ? 8282 : 8082);
    }
    if (!urlBuilder) {
        urlBuilder = basicUrlBuilder;
    }
    var agentClass = null;
    var agentOptions = null;
    if (socksHost) {
        var socksLocation = plywood_base_api_1.hostToLocation(socksHost, 1080);
        agentClass = secure ? SecureAgent : PlainAgent;
        agentOptions = {
            socksHost: socksLocation.hostname,
            socksPort: socksLocation.port
        };
        if (parameters.socksUsername)
            agentOptions.socksUsername = parameters.socksUsername;
        if (parameters.socksPassword)
            agentOptions.socksPassword = parameters.socksPassword;
    }
    function requestOptionsWithDecoration(opt) {
        return Promise.resolve()
            .then(function () {
            var query = opt.query, context = opt.context, options = opt.options;
            if (agentClass) {
                options.agentClass = agentClass;
                options.agentOptions = agentOptions;
            }
            if (secure) {
                options.strictSSL = (protocol === 'tls');
                if (ca)
                    options.ca = ca;
            }
            options.headers = options.headers || {};
            if (authToken && authToken.type === 'basic') {
                options.headers["Authorization"] = "Basic " + new Buffer(authToken.token).toString('base64');
            }
            if (requestDecorator) {
                var decorationPromise = requestDecorator({
                    method: options.method,
                    url: options.url,
                    query: JSON.parse(JSON.stringify(query))
                }, context['decoratorContext']);
                if (decorationPromise) {
                    return Promise.resolve(decorationPromise)
                        .then(function (decoration) {
                        if (!decoration)
                            return options;
                        if (decoration.method) {
                            options.method = decoration.method;
                        }
                        if (decoration.url) {
                            options.url = decoration.url;
                        }
                        if (decoration.headers) {
                            Object.assign(options.headers, decoration.headers);
                        }
                        if (decoration.query) {
                            if (typeof decoration.query === 'string') {
                                options.body = decoration.query;
                            }
                            else {
                                options.body = JSON.stringify(decoration.query);
                            }
                        }
                        if (decoration.resultType) {
                            options.resultType = decoration.resultType;
                        }
                        return options;
                    });
                }
            }
            return options;
        });
    }
    function requestPromiseWithDecoration(opt) {
        return requestOptionsWithDecoration(opt).then(requestPromise);
    }
    function failIfNoDatasource(url, query, timeout) {
        return requestPromiseWithDecoration({
            query: { queryType: "sourceList" },
            context: {},
            options: {
                method: "GET",
                url: url + "/druid/v2/datasources",
                json: true,
                timeout: timeout
            }
        })
            .then(function (resp) {
            if (getDataSourcesFromQuery(query).every(function (dataSource) { return resp.indexOf(dataSource) < 0; })) {
                throw new Error("No such datasource");
            }
            return null;
        });
    }
    return function (req) {
        var context = req.context || {};
        var query = req.query;
        var queryType = query.queryType, intervals = query.intervals;
        if (!queryType && typeof query.query === 'string') {
            queryType = 'sql';
        }
        var stream = new readable_stream_1.PassThrough({
            objectMode: true
        });
        if (intervals && (intervals === "1000-01-01/1000-01-02" || !intervals.length)) {
            process.nextTick(function () {
                stream.push(null);
            });
            return stream;
        }
        function streamError(e) {
            e.query = query;
            stream.emit('error', e);
            stream.end();
        }
        var url;
        locator()
            .then(function (location) {
            url = urlBuilder(location, secure);
            if (queryType === "status") {
                requestPromiseWithDecoration({
                    query: query,
                    context: context,
                    options: {
                        method: "GET",
                        url: url + '/status',
                        json: true,
                        timeout: timeout
                    }
                })
                    .then(function (resp) {
                    stream.push(resp);
                    stream.push(null);
                }, streamError);
                return;
            }
            if (queryType === "introspect" || queryType === "sourceList") {
                requestPromiseWithDecoration({
                    query: query,
                    context: context,
                    options: {
                        method: "GET",
                        url: url + "/druid/v2/datasources/" + (queryType === "introspect" ? getDataSourcesFromQuery(query)[0] : ''),
                        json: true,
                        timeout: timeout
                    }
                })
                    .then(function (resp) {
                    if (queryType === "introspect") {
                        if (Array.isArray(resp.dimensions) && !resp.dimensions.length &&
                            Array.isArray(resp.metrics) && !resp.metrics.length) {
                            return failIfNoDatasource(url, query, timeout).then(function () {
                                var err = new Error("Can not use GET route, data is probably in a real-time node or more than a two weeks old. Try segmentMetadata instead.");
                                err.query = query;
                                throw err;
                            });
                        }
                    }
                    return resp;
                })
                    .then(function (resp) {
                    stream.push(resp);
                    stream.push(null);
                }, streamError);
                return;
            }
            if (timeout != null) {
                query.context || (query.context = {});
                query.context.timeout = timeout;
            }
            requestOptionsWithDecoration({
                query: query,
                context: context,
                options: {
                    method: "POST",
                    url: url + "/druid/v2/" + (queryType === 'sql' ? 'sql/' : '') + (context['pretty'] ? '?pretty' : ''),
                    body: JSON.stringify(query),
                    headers: {
                        "Content-type": "application/json"
                    },
                    timeout: timeout
                }
            })
                .then(function (options) {
                request(options)
                    .on('error', function (err) {
                    if (err.message === 'ETIMEDOUT' || err.message === 'ESOCKETTIMEDOUT')
                        err = new Error("timeout");
                    streamError(err);
                })
                    .on('response', function (response) {
                    if (response.statusCode !== 200) {
                        response.on('error', streamError);
                        response.pipe(concat(function (resp) {
                            resp = String(resp);
                            var error;
                            try {
                                var body = JSON.parse(resp);
                                if (body && body.error === "Query timeout") {
                                    error = new Error("timeout");
                                }
                                else {
                                    var message = void 0;
                                    if (body && typeof body.error === 'string') {
                                        message = body.error;
                                        if (typeof body.errorMessage === 'string') {
                                            message = message + ": " + body.errorMessage;
                                        }
                                    }
                                    else {
                                        message = "Bad status code (" + response.statusCode + ")";
                                    }
                                    error = new Error(message);
                                    error.query = query;
                                    if (body && typeof body.host === 'string')
                                        error.host = body.host;
                                }
                            }
                            catch (e) {
                                error = new Error("bad response");
                            }
                            streamError(error);
                        }));
                        return;
                    }
                    var rowBuilder = new rowBuilder_1.RowBuilder({
                        resultType: options.resultType || queryType,
                        timestamp: hasOwnProperty(context, 'timestamp') ? context['timestamp'] : 'timestamp',
                        ignorePrefix: context['ignorePrefix'],
                        dummyPrefix: context['dummyPrefix']
                    });
                    rowBuilder.on('meta', function (meta) {
                        stream.emit('meta', meta);
                    });
                    rowBuilder.on('end', function () {
                        if (!rowBuilder.maybeNoDataSource) {
                            stream.end();
                            return;
                        }
                        failIfNoDatasource(url, query, timeout)
                            .then(function () {
                            stream.end();
                        }, streamError);
                    });
                    response
                        .pipe(new Combo({ packKeys: true, packStrings: true, packNumbers: true }))
                        .pipe(rowBuilder)
                        .pipe(stream, { end: false });
                });
            }, streamError);
        });
        return stream;
    };
}
exports.druidRequesterFactory = druidRequesterFactory;
