// chronoshift/build/timezone/timezone.js
import * as moment from 'moment-timezone';
var check;
var Timezone = (function () {
    function Timezone(timezone) {
        if (typeof timezone !== 'string') {
            throw new TypeError("timezone description must be a string");
        }
        if (timezone !== 'Etc/UTC' && !moment.tz.zone(timezone)) {
            throw new Error("timezone '" + timezone + "' does not exist");
        }
        this.timezone = timezone;
    }
    Timezone.isTimezone = function (candidate) {
        return candidate instanceof Timezone;
    };
    Timezone.formatDateWithTimezone = function (d, timezone) {
        var str;
        if (timezone && !timezone.isUTC()) {
            str = moment.tz(d, timezone.toString()).format('YYYY-MM-DDTHH:mm:ss.SSSZ');
        }
        else {
            str = d.toISOString();
        }
        return str.replace('.000', '');
    };
    Timezone.fromJS = function (spec) {
        return new Timezone(spec);
    };
    Timezone.prototype.valueOf = function () {
        return this.timezone;
    };
    Timezone.prototype.toJS = function () {
        return this.timezone;
    };
    Timezone.prototype.toJSON = function () {
        return this.timezone;
    };
    Timezone.prototype.toString = function () {
        return this.timezone;
    };
    Timezone.prototype.equals = function (other) {
        return Timezone.isTimezone(other) &&
            this.timezone === other.timezone;
    };
    Timezone.prototype.isUTC = function () {
        return this.timezone === 'Etc/UTC';
    };
    return Timezone;
}());
export { Timezone };
Timezone.UTC = new Timezone('Etc/UTC');
check = Timezone;
