// plyql/build/requester.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var plywood_1 = require("plywood");
var plywood_druid_requester_1 = require("plywood-druid-requester");
function properDruidRequesterFactory(options) {
    var druidHost = options.druidHost, retry = options.retry, timeout = options.timeout, verbose = options.verbose, concurrentLimit = options.concurrentLimit, requestDecorator = options.requestDecorator, socksHost = options.socksHost, socksUsername = options.socksUsername, socksPassword = options.socksPassword;
    var druidRequester = plywood_druid_requester_1.druidRequesterFactory({
        host: druidHost,
        timeout: timeout || 30000,
        requestDecorator: requestDecorator,
        socksHost: socksHost,
        socksUsername: socksUsername,
        socksPassword: socksPassword
    });
    if (retry) {
        druidRequester = plywood_1.retryRequesterFactory({
            requester: druidRequester,
            retry: retry,
            delay: 500,
            retryOnTimeout: false
        });
    }
    if (verbose) {
        druidRequester = plywood_1.verboseRequesterFactory({
            requester: druidRequester
        });
    }
    if (concurrentLimit) {
        druidRequester = plywood_1.concurrentLimitRequesterFactory({
            requester: druidRequester,
            concurrentLimit: concurrentLimit
        });
    }
    return druidRequester;
}
exports.properDruidRequesterFactory = properDruidRequesterFactory;
